worker/dbaccessor/tracker.go

```diff
@@ -0,0 +1,244 @@
+func (w *trackedDBWorker) DB(fn func(*sql.DB) error) error {
+	w.mutex.RLock()
+	// We have a fatal error, the DB can not be accessed.
+	if w.err != nil {
+		w.mutex.RUnlock()
+		return errors.Trace(w.err)
+	}
+	db := w.db
+	w.mutex.RUnlock()
+
+	return fn(db)
+}
+
+// Txn closes over a raw *sql.Tx. This allows retry semantics in only one
+// location. For instances where the underlying sql database is busy or if
+// it's a common retryable error that can be handled cleanly in one place.
+func (w *trackedDBWorker) Txn(ctx context.Context, fn func(context.Context, *sql.Tx) error) error {
+	return w.DB(func(db *sql.DB) error {
+		return database.Retry(ctx, func() error {
+			return database.Txn(ctx, db, fn)
+		})
+	})
+}
+
+// Err will return any fatal errors that have occurred on the worker, trying
+// to acquire the database.
+func (w *trackedDBWorker) Err() error {
+	w.mutex.RLock()
+	defer w.mutex.RUnlock()
+
+	return w.err
+}
+
+// Kill implements worker.Worker
+func (w *trackedDBWorker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker
+func (w *trackedDBWorker) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *trackedDBWorker) loop() error {
+	timer := w.clock.NewTimer(PollInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-timer.Chan():
+			// Any retryable errors are handled at the txn level. If we get an
+			// error returning here, we've either exhausted the number of
+			// retries or the error was fatal.
+			w.mutex.RLock()
+			currentDB := w.db
+			w.mutex.RUnlock()
+
+			newDB, err := w.verifyDB(currentDB)
+			if err != nil {
+				// If we get an error, ensure we close the underlying db and
+				// mark the tracked db in an error state.
+				w.mutex.Lock()
+				if err := w.db.Close(); err != nil {
+					w.logger.Errorf("error closing database: %v", err)
+				}
+				w.err = errors.Trace(err)
+				w.mutex.Unlock()
+
+				// As we failed attempting to verify the db, we're in a fatal
+				// state. Collapse the worker and if required, cause the other
+				// workers to follow suite.
+				return errors.Trace(err)
+			}
+
+			// We've got a new DB. Close the old one and replace it with the
+			// new one, if they're not the same.
+			if newDB != currentDB {
+				w.mutex.Lock()
+				if err := w.db.Close(); err != nil {
+					w.logger.Errorf("error closing database: %v", err)
+				}
+				w.db = newDB
+				w.err = nil
+				w.mutex.Unlock()
+			}
+		}
+	}
+}
+
+func (w *trackedDBWorker) verifyDB(db *sql.DB) (*sql.DB, error) {
+	// Force the timeout to be lower that the DefaultTimeout,
+	// so we can spot issues sooner.
+	// Also allow killing the tomb to cancel the context,
+	// so shutdown/restart can not be blocked by this call.
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*10)
+	ctx = w.tomb.Context(ctx)
+	defer cancel()
+
+	if w.logger.IsTraceEnabled() {
+		w.logger.Tracef("verifying database by attempting a ping")
```

Comment:
I noticed this elsewhere, do we have any additional context that we could pass here? attemping a ping against IP or something like that?

Reply:
We don't have the raw instance information, just the *sql.DB. I'll see what's available to us.